src/main/java/ch/jalu/configme/migration/VersionMigrationService.java

```diff
@@ -0,0 +1,236 @@
+        this.versionPropertyName = versionPropertyName;
+        this.migrations = Collections.unmodifiableCollection(migrations);
+    }
+
+    @Override
+    public boolean checkAndMigrate(@NotNull PropertyReader reader, @NotNull ConfigurationData configurationData) {
+        int currentConfigVersion = readConfigurationVersion(reader, configurationData);
+
+        if (currentConfigVersion < getStartVersion()) handleTooLowerVersion(currentConfigVersion);
+        if (currentConfigVersion > getCurrentVersion()) handleTooHigherVersion(currentConfigVersion);
+
+        return performMigrations(reader, configurationData) || !configurationData.areAllValuesValidInResource();
+    }
+
+    /**
+     * @return the start config version. By default, it is 1.
+     */
+    public int getStartVersion() {
+        return this.startVersion;
+    }
+
+    /**
+     * @return the current config version.
+     */
+    public int getCurrentVersion() {
+        return this.currentVersion;
+    }
+
+    /**
+     * @return The config property name. By default, it is "version".
+     */
+    @NotNull
+    public String getVersionPropertyName() {
+        return this.versionPropertyName;
+    }
+
+    /**
+     * @return The unmodifiable {@link Collection<Migration>} of migrations.
+     */
+    @NotNull
+    public Collection<Migration> getMigrations() {
+        return this.migrations;
+    }
+
+    /**
+     * Invoked if the version is lower than {@link #getStartVersion()}.
+     */
+    protected void handleTooLowerVersion(int version) {
+        throw new IllegalStateException("The current version (" + version + ") is lower than " + getStartVersion());
+    }
+
+    /**
+     * Invoked if the version is greater than {@link #getCurrentVersion()}.
+     */
+    protected void handleTooHigherVersion(int version) {
+        throw new IllegalStateException("The current version (" + version + ") is greater than " + getCurrentVersion());
+    }
+
+    /**
+     * Perform the migration by using the versioning system.
+     * <p>
+     * Note that the settings manager automatically saves the resource
+     * if the migration service returns {@link #MIGRATION_REQUIRED} from {@link #checkAndMigrate}.
+     *
+     * @param reader the reader with which the configuration file can be read
+     * @param configurationData the configuration data
+     * @return true if a migration has been performed, false otherwise (see constants on {@link MigrationService})
+     */
+    protected boolean performMigrations(@NotNull PropertyReader reader, @NotNull ConfigurationData configurationData) {
+        int currentConfigVersion = readConfigurationVersion(reader, configurationData);
+        AtomicBoolean migrationResult = new AtomicBoolean(NO_MIGRATION_NEEDED);
+
+        // Migrate the configuration from the version 1 to 2 to 3 to ...
+        migrations
+            .forEach(migration -> {
+                if (migration == null) return; // null-safe check.
```

Comment:
Not sure this is necessary. If you want to keep it, please add curly braces :smile: